src/python/pants/backend/makeself/makeself.py

```diff
@@ -0,0 +1,322 @@
+    )
+
+
+class MakeselfDistribution(DownloadedExternalTool):
+    """The makeself distribution.
+
+    You can find releases here: https://github.com/megastep/makeself/releases
+    """
+
+
+@rule(desc="Download makeself distribution", level=LogLevel.DEBUG)
+async def download_makeself_distribution(
+    options: MakeselfSubsystem,
+    platform: Platform,
+) -> MakeselfDistribution:
+    tool = await Get(
+        DownloadedExternalTool,
+        ExternalToolRequest,
+        options.get_request(platform),
+    )
+    logger.debug("makeself external tool: %s", tool)
+    return MakeselfDistribution(digest=tool.digest, exe=tool.exe)
+
+
+class MakeselfTool(DownloadedExternalTool):
+    """The Makeself tool."""
+
+
+@rule(desc="Extract makeself distribution", level=LogLevel.DEBUG)
+async def extract_makeself_distribution(
+    dist: MakeselfDistribution,
+) -> MakeselfTool:
+    out = "__makeself"
+    result = await Get(
+        ProcessResult,
+        RunMakeselfArchive(
+            exe=dist.exe,
+            input_digest=dist.digest,
+            output_directory=out,
+            description=f"Extracting Makeself archive: {out}",
+            level=LogLevel.DEBUG,
+        ),
+    )
+    digest = await Get(Digest, RemovePrefix(result.output_digest, out))
+    return MakeselfTool(digest=digest, exe="makeself.sh")
+
+
+@dataclass(frozen=True)
+class CreateMakeselfArchive:
+    archive_dir: str
+    file_name: str
+    label: str
+    startup_script: str
+    input_digest: Digest
+    description: str = dataclasses.field(compare=False)
+    output_filename: str
+    level: LogLevel = LogLevel.INFO
+    cache_scope: Optional[ProcessCacheScope] = None
+    timeout_seconds: Optional[int] = None
+
+
+@rule
+async def create_makeself_archive(
+    request: CreateMakeselfArchive,
+    makeself: MakeselfTool,
+    awk: AwkBinary,
+    basename: BasenameBinary,
+    cat: CatBinary,
+    date: DateBinary,
+    dirname: DirnameBinary,
+    du: DuBinary,
+    expr: ExprBinary,
+    find: FindBinary,
+    gzip: GzipBinary,
+    rm: RmBinary,
+    sed: SedBinary,
+    sh: ShBinary,
+    sort: SortBinary,
+    tar: TarBinary,
+    wc: WcBinary,
+    xargs: XargsBinary,
+    tr: TrBinary,
+    cksum: CksumBinary,
+    cut: CutBinary,
+    chmod: ChmodBinary,
+) -> Process:
+    shims = await Get(
+        BinaryShims,
+        BinaryShimsRequest(
+            paths=(
+                awk,
+                basename,
+                cat,
+                date,
+                dirname,
+                du,
+                expr,
+                find,
+                gzip,
+                rm,
+                sed,
+                sh,
+                sort,
+                tar,
+                wc,
+                tr,
+                cksum,
+                cut,
+                chmod,
+                xargs,
+            ),
+            rationale="create makeself archive",
+        ),
+    )
+    tooldir = "__makeself"
```

Comment:
Could this be either moved to a global or placed in `MakeselfTool` (potentially have the `exe` field be the full path, including the `__makeself`?), so that it's not duplicating the `out = ...` in `extract_makeself_distribution`?